cache_archive.go

```diff
@@ -0,0 +1,252 @@
+		info, err := os.Lstat(pth)
+
+		var link string
+		if info.Mode()&os.ModeSymlink != 0 {
+			link, err = os.Readlink(pth)
+			if err != nil {
+				return err
+			}
+		}
+
+		header, err := tar.FileInfoHeader(info, link)
+		if err != nil {
+			return err
+		}
+
+		header.Name = pth
+		header.ModTime = info.ModTime()
+
+		if err := a.tar.WriteHeader(header); err != nil {
+			return err
+		}
+
+		file, err := os.Open(pth)
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			if err := file.Close(); err != nil {
+				log.Warnf("Failed to close file (%s): %s", pth, err)
+			}
+		}()
+
+		if _, err := io.CopyN(a.tar, file, info.Size()); err != nil && err != io.EOF {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WriteHeader writes the cache descriptor file into the archive as a tar header.
+func (a *Archive) WriteHeader(descriptor map[string]string, descriptorPth string) error {
+	b, err := json.Marshal(descriptor)
+	if err != nil {
+		return err
+	}
+
+	closingHeader := &tar.Header{
+		Name:     descriptorPth,
+		Size:     int64(len(b)),
+		Typeflag: tar.TypeReg,
+		Mode:     0600,
+		ModTime:  time.Now(),
+	}
+
+	if err := a.tar.WriteHeader(closingHeader); err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(a.tar, bytes.NewReader(b)); err != nil && err != io.EOF {
+		return err
+	}
+	return nil
+}
+
+// Close closes the archive.
+func (a *Archive) Close() error {
+	if err := a.tar.Close(); err != nil {
+		return err
+	}
+
+	if a.gzip != nil {
+		if err := a.gzip.Close(); err != nil {
+			return err
+		}
+	}
+
+	return a.file.Close()
+}
+
+// uploadArchive uploads the archive file to a given destination.
+// If the destination is a local file path (url has a file:// scheme) this function copies the cache archive file to the destination.
+// Otherwise destination should point to the Bitrise cache API server, in this case the function has builtin retry logic with 3s sleep.
+func uploadArchive(pth, url string) error {
+	fi, err := os.Stat(pth)
+	if err != nil {
+		return fmt.Errorf("failed to get file info (%s): %s", pth, err)
+	}
+	sizeInBytes := fi.Size()
+	log.Printf("Archive file size: %d bytes / %f MB", sizeInBytes, (float64(sizeInBytes) / 1024.0 / 1024.0))
+
+	uploadURL, err := getCacheUploadURL(url, sizeInBytes)
+	if err != nil {
+		return fmt.Errorf("failed to generate upload url: %s", err)
+	}
+
+	if err := tryToUploadArchive(uploadURL, pth); err != nil {
+		fmt.Println()
+		log.Warnf("First upload attempt failed, retrying...")
+		fmt.Println()
+		time.Sleep(3000 * time.Millisecond)
+		return tryToUploadArchive(uploadURL, pth)
+	}
+	return nil
+}
+
+// getCacheUploadURL requests an upload url from the Bitrise cache API server.
+func getCacheUploadURL(cacheAPIURL string, fileSizeInBytes int64) (string, error) {
+	if strings.HasPrefix(cacheAPIURL, "file://") {
+		return cacheAPIURL, nil
+	}
+
+	req, err := http.NewRequest("POST", cacheAPIURL, bytes.NewReader([]byte(fmt.Sprintf(`{"file_size_in_bytes": %d}`, fileSizeInBytes))))
```

Comment:
You could use `http.MethodPost` instead of `"POST"` string.
Ofc. both are the same, but If you need to change something else in the PR it would be nice to change this as well.